src/transfer.cpp

```diff
@@ -1192,8 +1213,21 @@ void DirectReadNode::cmdresult(error e, dstime timeleft)
         // feed all pending reads to the global read queue
         for (dr_list::iterator it = reads.begin(); it != reads.end(); it++)
         {
-            assert((*it)->drq_it == client->drq.end());
-            (*it)->drq_it = client->drq.insert(client->drq.end(), *it);
+            DirectRead* dr = *it;
+            assert(dr->drq_it == client->drq.end());
+
+            if (dr->drbuf.tempUrlVector().empty())
+            {
+                // DirectRead starting
+                dr->drbuf.setIsRaid(dr->drn->tempurls, dr->offset, dr->offset + dr->count, dr->drn->size, 2097152);  // 2 MB max buffer usage approx for streaming
+            }
+            else
+            {
+                // URLs have been re-requested, eg. due to temp URL expiry.  Keep any parts downloaded already
+                dr->drbuf.updateUrlsAndResetPos(dr->drn->tempurls);
```

Comment:
This is the only place in the code where URLs are updated (by keeping downloaded parts). I wonder if there's any other codepath where it should be done :thinking:

Reply:
For transfers a similar thing can happen when a transfer is rescheduled when the known tempurls have been cleared (eg. after a certain number of errors, or the URLs are too old after loading from cache).   I've kept the same logic for exactly when we fetch new URLs, except that they are stored in the RaidBufferManager now (to abstract away the use of 1 or 6 URLs).  Hopefully that makes it a bit easier to review, by keeping changes to a minimum.  thanks